### ISSUE TITLE
gradient blows up for linear regression with standard SGD

### ISSUE THREAD

**Issue by dustinvtran** [airoldilab/sgd#41]
Implicit works but this one doesn't (it should)
```{R}
library(sgd)

# Dimensions
N <- 1e5
d <- 1e2

# Generate data.
X <- matrix(rnorm(N*d), ncol=d)
theta <- rep(5, d+1)
eps <- rnorm(N)
y <- cbind(1, X) %*% theta + eps
dat <- data.frame(y=y, x=X)

sgd.theta <- sgd(y ~ ., data=dat, model="lm", sgd.control=list(method="sgd"))
## NA or infinite gradient
## Error in fit(x, y, model.control, sgd.control) :
##   An error has occured, program stopped.
```

**Comment by dustinvtran** [comment]
Can't seem to find the issue without going deep into the C++ code.. Oh well, I am submitting to CRAN without fixing this first.

**Comment by ptoulis** [comment]
Is it possible that the package will be denied because of that?

**Comment by dustinvtran** [comment]
I'm not quite sure. It seems CRAN looks more into formatting and basic example cases. I don't think they have enough expert knowledge to know/care about robustness or properties of non-examples. Will submit and let you all know what happens when they reply

**Comment by ptoulis** [comment]
Ok let's try then.

is there an option to mark the package as experimental?

**Comment by lantian2012** [comment]
I think CRAN just checks format. I'm not sure they manually checks something. But at least, the package needs to give no warning when we run CMD CHECK on it.

**Comment by ptoulis** [comment]
was the package submitted?

when will it be possible to find it online ?
(would like to link to it in the presentation)

**Comment by dustinvtran** [comment]
From CRAN

> The mandatory ‘Version’ field gives the version of the package. This is a sequence of at least two (and usually three) non-negative integers separated by single ‘.’ or ‘-’ characters. The canonical form is as shown in the example, and a version such as ‘0.01’ or ‘0.01.0’ will be handled as if it were ‘0.1-0’. It is not a decimal number, so for example 0.9 < 0.75 since 9 < 75.

So unfortunately I can't add the term alpha/beta to CRAN, although I added "alpha" to the version release on github. It's pending submission for CRAN at the moment and will take at least a few days. Meanwhile you can still point them to https://github.com/airoldilab/sgd/releases if need be :)

**Comment by ptoulis** [comment]
It's ok. i will just name the package so people are aware.

thanks for the swift actions Dustin!

**Comment by dustinvtran** [comment]
:) good luck tomorrow!

**Comment by dustinvtran** [comment]
We need to remove `transfer.h` to simplify the gradient

**Comment by lantian2012** [comment]
That file was added to store derivatives information for glm. In the current implementation, it is dynamically bound to an instance of `Sgd_Experiment_Glm `. It is not required for the more general `Sgd_Experiment`. 
I think we may rename it, and use that file to store model specific functions.

**Comment by dustinvtran** [comment]
Turns out the issue was with how the `one-dim` learning rate works. Using the minimum eigenvalue of the covariance of the data in order to set the hyperparameter alpha leads to `1e-32` in the linear regression example. This causes the learning rate to just be the constant 1.

Implicit of course still was able to converge so it was hard to spot that error. :-)

**Comment by dustinvtran** [comment]
Still need to find a better way to set the hyperparameter alpha.

**Comment by dustinvtran** [comment]
I'll add the need to set the hyperparameter alpha in a separate (lower priority) issue, as this blowup is fixed
